aws/protocol/ec2query/error_utils.go

```diff
@@ -0,0 +1,28 @@
+package ec2query
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+)
+
+// errorResponse denotes the error response structure for ec2Query error response
+type errorResponse struct {
+	Code    string `xml:"Errors>Error>Code"`
+	Message string `xml:"Errors>Error>Message"`
+}
+
+// GetEc2QueryResponseErrorCode returns the error code, error message from an ec2query error response body
+func GetResponseErrorCode(r io.Reader) (code string, message string, err error) {
+	rb, err := ioutil.ReadAll(r)
+	if err != nil {
+		return "", "", err
+	}
+
+	var er errorResponse
+	if err := xml.Unmarshal(rb, &er); err != nil {
```

Comment:
What do you think about using [xml.NewDecoder](https://pkg.go.dev/encoding/xml?tab=doc#NewDecoder) instead of Unmarshal?  This will prevent allocating an additional buffer for the error message.

Reply:
I was hoping to make a decision on this when we do benchmarking. How would using xml.NewDecoder save an allocation here? I chose Unmarshal as we have a defined structure for modeled errors.

Reply:
The allocation will happen when creating the `rb` buffer via the `ioutil.ReadAll` call. This allocation will be in addition to any other allocations created by the xml Unmarshaler's internal behavior. `xml.Decoder` will stream the bytes in without creating the additional byte slice buffer for the overall contents.